internal/oauth2/lazy_token.go

```diff
@@ -0,0 +1,119 @@
+package oauth2
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+// Tokener is an interface that defines the Token method.
+type Tokener interface {
+	Token(context.Context) (string, error)
+}
+
+// LazyToken is a struct that handles token acquisition and caching.
+type LazyToken struct {
+	url          string
+	clientId     string
+	clientSecret string
+	token        string
+	offlineToken string
+	expiration   time.Time
+	mutex        sync.Mutex
+}
+
+type tokenResponse struct {
+	AccessToken string    `json:"access_token"`
+	Expiration  time.Time `json:"expiration"`
+}
+
+// NewLazyToken creates a new LazyToken instance.
+func NewLazyToken(tokenURL, clientId, clientSecret string) *LazyToken {
+	return &LazyToken{
+		url:          tokenURL,
+		clientId:     clientId,
+		clientSecret: clientSecret,
+		token:        "",
+		expiration:   time.Now(),
+	}
+}
```

Comment:
Might be useful to create also `NewXXX(url, id, existingToken)`

Reply:
why the time.Now() is not good enough?

Reply:
I think we agreed on actually removing `New` method after all. Struct can be initialized normally, expiration time will default to year 0001 which is in the past (expired) so that is good.

Reply:
I initialized it at the client itslef -
rc := RecommendationsClient{
		URL: conf.URL,
		Tokener: &oauth2.LazyToken{
			Url:          "",
			ClientId:     "",
			ClientSecret: "",
			Token:        "",
			Expiration:   time.Time{},
		},
		client: client,
	}
should I do it in the lazy_token?